apps/system/js/sync_manager.js

```diff
@@ -0,0 +1,547 @@
+
+      this.error = error;
+
+      if (SyncUnrecoverableErrors.indexOf(error) > -1) {
+        this.debug('Unrecoverable error');
+        Service.request('SyncStateMachine:disable');
+      }
+    },
+
+    _handle_onsyncsyncing: function() {
+      this.debug('onsyncsyncing observed');
+      this.saveState();
+
+      // In case that we are syncing on user demand, we unregister the
+      // sync request.
+      // It's harmless to unregister an expired request, so it's also ok
+      // to do it for a periodic sync.
+      this.unregisterSyncRequest();
+
+      var collections = [];
+      COLLECTIONS.forEach(name => {
+        if (this.settings.get('sync.collection.' + name + '.enabled')) {
+          collections.push(name);
+        }
+      });
+      if (!collections.length) {
+        Service.request('SyncStateMachine:success');
+        return;
+      }
+
+      var args = [];
+      this.getAssertion().then(assertion => {
+        args.push(assertion);
+        return this.getKeys();
+      }).then(keys => {
+        args.push(keys);
+        args.push(collections);
+        this.doSync.apply(this, args);
+      }).catch(error => {
+        console.error('Sync error', error);
+        Service.request('SyncStateMachine:error', ERROR_SYNC_REQUEST);
+      });
+    },
+
+    _handle_fxaEvent: function(event) {
+      if (!event || !event.detail) {
+        console.error('Wrong event');
+        return;
+      }
+
+      if (event.detail.eventName != 'onlogout') {
+        return;
+      }
+
+      // Logging out from Firefox Accounts should disable Sync.
+      Service.request('SyncStateMachine:disable');
+    },
+
+    _handle_killapp: function(event) {
+      // The synchronizer app can be killed while it's handling a sync
+      // request. In that case, we need to record an error and notify
+      // the state machine about it. Otherwise we could end up on a
+      // permanent 'syncing' state.
+      var killedApp = event.detail.origin;
+      navigator.mozApps.getSelf().onsuccess = event => {
+        var app = event.target.result;
+        app.getConnections().then(connections => {
+          connections.forEach(connection => {
+            if (connection.keyword != SYNC_REQUEST_IAC_KEYWORD ||
+                connection.subscriber != killedApp) {
+              return;
+            }
+            Service.request('SyncStateMachine:error', ERROR_SYNC_APP_KILLED);
+          });
+
+        });
+      };
+    },
+
+    /** Helpers **/
+
+    initStore: function() {
+      var promises = [];
+      [SYNC_STATE,
+       SYNC_STATE_ERROR,
+       SYNC_LAST_TIME].forEach(key => {
+        var promise = new Promise(resolve => {
+          asyncStorage.getItem(key, value => {
+            this.debug(key + ': ' + value);
+            this.store.set(key, value);
+            resolve();
+          });
+        });
+        promises.push(promise);
+      });
+      return Promise.all(promises);
+    },
+
+    saveState: function(state) {
```

Comment:
nit: I'd rename to `updateState` since the saving is done by the property setter and it fits nicely with the fact that without parameter this will do a refresh.